### PR TITLE
Add constant time fasta index seeking [NEED REVIEW]

### PR DESCRIPTION
**Pull request by jakobnissen**
Addresses issue #29. This PR also includes #30 .

Changes:
* `FASTA.Index` now uses a `Dict` instead of a vector of names. This is slower to construct and more memory inefficient, but obviously has the benefit of constant-time lookup in the index. I'd argue that if you create an index, you do so specifically in order to have fast lookup, and you probably don't care about the few extra kilobytes of RAM used, or miliseconds when constructing the index.
* Added a new method: `extract(r::Reader, A::Alphabet, s::AbstractString, u::UnitRange)`, which does the same as `sequence(A, r[s], u)`, but does not load the entire sequence into memory first.

### CODE REVIEW

File: src/fasta/reader.jl [hunk 1]
```diff
@@ -71,6 +71,35 @@ function Base.getindex(reader::Reader, name::AbstractString)
...
+    seek(reader.state.stream, offset + (first(range) - 1) + (len_newline * skip_lines))
+
+    # Now fill in data
+    buffer = Vector{UInt8}(undef, length(range))
+    filled = 0
+    while filled < length(buffer)
+        line = readline(reader.state.stream)
+        len = min(ncodeunits(line), length(buffer) - filled)
+        unsafe_copyto!(pointer(buffer, filled+1), pointer(line), len)
+        filled += len
```
Comment (CiaranOMara) [inline src/fasta/reader.jl:98]: Counting down would remove the calculation for `len`.
```julia
remaining = length(range)
while remaining > linebase
    line = readline(reader.state.stream)
    remaining -= write(buffer, line)
end
line = read(reader.state.stream, remaining)
write(buffer, line)
```

I don't know whether `unsafe_copyto!` is faster than `write` here.

File: src/fasta/index.jl [hunk 2]
```diff
@@ -41,30 +41,31 @@ function read_faidx(input::IO)
...
+    # sequence in bytes, then seek to right after that one.
+    else
+        prev_offset = index.offsets[i - 1]
+        prev_len = index.lengths[i - 1]
+        prev_linebase = index.linebases[i - 1]
+        prev_linewidth = index.linewidths[i - 1]
+
+        newline_len = prev_linewidth - prev_linebase
+        len = cld(prev_len, prev_linebase) * newline_len + prev_len
+        offset = prev_offset + len
```
[the remark this answers is not in the document]
Comment (CiaranOMara) [reply src/fasta/index.jl:67]: The number of bytes used for the line termination character, could be calculated once in `read_faidx` with something like the following, then added to `Index`.
```julia
line = readuntil(s, 0x0a, keep = true)
terminator_ncodeunits = line[i-1] == 0x0d ? 2 : 1
```
ref: https://github.com/JuliaLang/julia/blob/697e782ab86bfcdd7fd15550241fe162c51d9f98/base/io.jl#L482-L492

[the remark this answers is not in the document]
Comment (jakobnissen) [reply src/fasta/index.jl:67]: Alright, then the logic is sound. Unfortunately, the spec allows differing line termination lengths for different sequences in the same file. Hence, we can't just calculate it once. (one could argue allowing that is insanity, I guess). Nonetheless here, the per-sequence calculation is easy.

File: src/fasta/index.jl [hunk 3]
```diff
@@ -41,30 +41,31 @@ function read_faidx(input::IO)
...
     end
-    n_back *= 2
-    @goto seekback
+    seek(input, offset)
+    return nothing
```
[the remark this answers is not in the document]
Comment (jakobnissen) [reply src/fasta/index.jl:70]: Agreed, but I'd rather not break it here. Perhaps the entire indexing could use an overhaul soon (e.g. there is no function to create an index).

File: src/fasta/index.jl [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 # http://www.htslib.org/doc/faidx.html
 struct Index
-    names::Vector{String}
+    names::Dict{String, Int}
     lengths::Vector{Int}
     offsets::Vector{Int}
```
[the remark this answers is not in the document]
Comment (jakobnissen) [reply src/fasta/index.jl:13]: We could add a comment, but the meaning of the fields is defined in the faidx doc, linked to in the comment above the struct definition

File: src/fasta/reader.jl (hunk 1 above)
Comment (jakobnissen) [reply src/fasta/reader.jl:98]: You're right - unsafe_coptyo! is a little faster, write would work for an IOBuffer. I think both would be fine :)

File: src/fasta/index.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (CiaranOMara) [reply src/fasta/index.jl:67]: > Unfortunately, the spec allows differing line termination lengths for different sequences in the same file.

You’re correct. I had missed that detail. That detail was in the last half of a sentence where they qualify consistency with non-consistency. 

> (one could argue allowing that is insanity, I guess)

I suppose there’s a case for splicing information from different machines.

**Comment by BenJWard** [comment]
Is this to be merged? I already merged the other related PR.

**Comment by CiaranOMara** [comment]
The new indexing logic proposed here looks good to me. Programatically this PR is good to merge.

Though maintenance-wise there was, however, one thing that wasn't crucially obvious to me. I got caught out not realising that the spec allows differing line termination lengths/characters for different sequences in the same file. My temptation was to optimise out what I had perceived to be duplicate calculations of the number of bytes used for the line termination character. This optimisation, of course, would have been wrong. I'd like to see a comment that warns against such a temptation and a test that ensures indexing works with differing line termination characters (I didn't see such a test when I skimmed `runtests.jl`).

**Comment by jakobnissen** [comment]
Added tests and comments as requested now. If CI gives me its blessings, this can be merged.

**Pull request merged**